presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -15,101 +15,124 @@
+    @GuardedBy("this")
+    private CompletableFuture<?> notFullSignal = new CompletableFuture<>();
+    // This future is completed when the queue transitions from empty to not. But it will be replaced by a new instance of future immediately.
+    @GuardedBy("this")
+    private CompletableFuture<?> notEmptySignal = new CompletableFuture<>();
+    @GuardedBy("this")
+    private boolean finishing = false;
+
+    private Executor executor;
+
+    public AsyncQueue(int targetQueueSize, Executor executor)
     {
-        requireNonNull(element, "element is null");
-        if (finished.get()) {
-            return;
-        }
-        queue.add(element);
-        notEmptyFuture.get().complete(null);
+        checkArgument(targetQueueSize >= 1, "targetQueueSize must be at least 1");
+        this.targetQueueSize = targetQueueSize;
+        this.elements = new ArrayDeque<>(targetQueueSize * 2);
+        this.executor = executor;
+    }
+
+    public synchronized int size()
+    {
+        return elements.size();
     }
 
-    /**
-     * Gets a batch of elements from the queue.  If queue is not empty, a
-     * completed future containing the queued elements is returned.  If the
-     * queue is empty, an incomplete future is returned that completes when
-     * an element is added.
-     * <p>
-     * It is possible that an empty list will be returned as another reader
-     * may have consumed all of the data.  A caller should read until, the
-     * {@code isFinished} method returns {@code true}.
-     *
-     * @param maxSize the maximum number of elements to return.
-     */
-    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    public synchronized boolean isFinished()
     {
-        return notEmptyFuture.get().thenApply(x -> getBatch(maxSize));
+        return finishing && elements.size() == 0;
     }
 
-    private List<T> getBatch(int maxSize)
+    public synchronized void finish()
     {
-        // take up to maxSize elements from the queue
-        List<Object> elements = new ArrayList<>(maxSize);
-        queue.drainTo(elements, maxSize);
-
-        // if the queue is empty and the current future is finished, create a new one so
-        // a new readers can be notified when the queue has elements to read
-        if (queue.isEmpty() && !finished.get()) {
-            CompletableFuture<?> future = notEmptyFuture.get();
-            if (future.isDone()) {
-                notEmptyFuture.compareAndSet(future, new CompletableFuture<>());
-            }
+        if (finishing) {
+            return;
         }
+        finishing = true;
 
-        // if someone added an element while we were updating the state, complete the current future
-        if (!queue.isEmpty() || finished.get()) {
-            notEmptyFuture.get().complete(null);
+        if (elements.size() == 0) {
+            CompletableFuture<?> notEmptySignal = this.notEmptySignal;
+            executor.execute(() -> notEmptySignal.complete(null));
```

Comment:
You could introduce a method to do this and you wouldn't need to capture the field in a local variable:


```java
if (elements.size() == 0) {
    completeAsync(executor, notEmptySignal);
}
```

```java
private static void completeAsync(Executor executor, CompletableFuture<?> future)
{
    executor.execute(() -> future.complete(null);
}
```